### PR TITLE
engine: container BaD's can handle image targets with LiveUpdate [ch2039]

### PR DESCRIPTION
**Pull request by maiamcc**
Hello @landism, @nicks,

Please review the following commits I made in branch maiamcc/targets-for-live-upd:

8442ab8c58bde4092991fc0a55fcdf9849ad5ad3 (2019-03-28 16:45:22 -0400)
synclet BaD can handle LiveUpdate / tests green

ee7e5c0272d16ff931bd51c7f4780ea534603ab0 (2019-03-28 16:35:48 -0400)
red tests for synclet BaD

19182300872febdfa6daf21a9e7302a3b53f1c93 (2019-03-28 15:04:05 -0400)
are these tests better or worse?

8d2d5536ac6d61a724b7c81253f0764591df2d84 (2019-03-28 14:09:59 -0400)
behavior implemented for localContainerBuild / green tests

153e06faf47ee133aeca3440fb973a6398aaed87 (2019-03-28 13:16:24 -0400)
red test for localContainerBaD

e367c3b2ebb7f0f3514bab75a6af04ad57f3dcbd (2019-03-28 13:04:04 -0400)
extract liveupdate targets

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

File: internal/engine/build_and_deployer_live_update_test.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/internal/synclet/sidecar"
+)
+
+type testCase struct {
+	env k8s.Env
+
+	baseManifest        model.Manifest
+	liveUpdSyncs        []model.LiveUpdateSyncStep
```
Comment (nicks) [inline internal/engine/build_and_deployer_live_update_test.go]: i'm surprised the test case doesn't take a LiveUpdate object? and then factor out this code into a helper that makes LiveUpdate objects

File: internal/engine/build_and_deployer_live_update_test.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	assert.Equal(t, tCase.expectSyncletUpdateContainerCount, f.sCli.UpdateContainerCount, "synclet update container")
+	f.assertContainerRestarts(tCase.expectDockerRestartCount)
+	assert.Equal(t, tCase.expectSyncletHotReload, f.sCli.UpdateContainerHotReload, "synclet hot reload")
+
+	id := manifest.ImageTargetAt(0).ID()
+	_, hasResult := result[id]
+	assert.True(t, hasResult)
+	assert.Equal(t, k8s.MagicTestContainerID, result.OneAndOnlyContainerID().String())
+
+	if tCase.expectK8sDeploy {
```
Comment (nicks) [inline internal/engine/build_and_deployer_live_update_test.go]: if this is false, shouldn't we assert somewhere that there was no deploy?

File: internal/engine/build_and_deployer_live_update_test.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+	expectSyncletHotReload            bool
+	expectK8sDeploy                   bool
+	expectSyncletDeploy               bool
+}
+
+func runTestCase(t *testing.T, f *bdFixture, tCase testCase) {
+	var bs store.BuildStateSet
+	if tCase.changedFile != "" {
+		changed := f.WriteFile(tCase.changedFile, "blah")
+		bs = resultToStateSet(alreadyBuiltSet, []string{changed}, f.deployInfo())
```
Comment (nicks) [inline internal/engine/build_and_deployer_live_update_test.go]: fwiw, the `alreadyBuiltSet` global you use here isn't really generalizable, it really only works with the sancho test data

Comment (maiamcc) [reply internal/engine/build_and_deployer_live_update_test.go]: yeah -- are you suggesting i change it, or just an fyi?

File: internal/model/live_update.go [hunk 4]
```diff
@@ -61,7 +79,50 @@ type LiveUpdateRunStep struct {
 
 func (l LiveUpdateRunStep) liveUpdateStep() {}
 
+func (l LiveUpdateRunStep) toRun() Run {
+	r := Run{Cmd: l.Command}
+	if l.Trigger != "" {
+		r = r.WithTriggers([]string{l.Trigger}) // TODO(maia): should be []string to match Run.Triggers
```
[the remark this answers is not in the document]
Comment (maiamcc) [reply internal/model/live_update.go]: yay!

File: internal/model/live_update.go [hunk 5]
```diff
@@ -61,7 +79,50 @@ type LiveUpdateRunStep struct {
...
+
 // Specifies that the container should be restarted when any files in `Sync` steps have changed.
 type LiveUpdateRestartContainerStep struct{}
 
 func (l LiveUpdateRestartContainerStep) liveUpdateStep() {}
+
+// TODO(maia): s/Mount/Sync
+func (lu LiveUpdate) SyncSteps() []Mount {
+	var syncs []Mount
+	for _, step := range lu.Steps {
```
[the remark this answers is not in the document]
Comment (maiamcc) [reply internal/model/live_update.go]: yyyyeah i think so! and populating base dir of the run steps?

File: internal/model/live_update.go [hunk 6]
```diff
@@ -61,7 +79,50 @@ type LiveUpdateRunStep struct {
...
+		switch step := step.(type) {
+		case LiveUpdateSyncStep:
+			syncs = append(syncs, step.toMount())
+		}
+	}
+	return syncs
+}
+
+func (lu LiveUpdate) RunSteps() []Run {
+	// TODO(maia): populate run.BaseDirectory with Workdir (if given)
```
[the remark this answers is not in the document]
Comment (maiamcc) [reply internal/model/live_update.go]: Already captured in [ch2043](https://app.clubhouse.io/windmill/story/2043/make-sure-run-s-exec-relative-to-workdir), no?

(also ugh fuck me.)

**Pull request merged**